compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/dfa/FirLocalVariableAssignmentAnalyzer.kt

```diff
@@ -0,0 +1,355 @@
+                // TODO: consider to also handle `throw`, which would require keeping track of all `try` and `catch` constructs.
+                data.flow = null
+            }
+
+            @OptIn(ExperimentalStdlibApi::class)
+            override fun visitFunctionCall(functionCall: FirFunctionCall, data: MiniCfgData) {
+                val visitor = this
+                with(functionCall) {
+                    setOfNotNull(explicitReceiver, dispatchReceiver, extensionReceiver).forEach { it.accept(visitor, data) }
+                    // Delay processing of lambda args because lambda body are evaluated after all arguments have been evaluated.
+                    val (lambdaArgs, normalArgs) = argumentList.arguments.partition { it is FirAnonymousFunction && it.isLambda }
+                    normalArgs.forEach { it.accept(visitor, data) }
+                    lambdaArgs.forEach { it.accept(visitor, data) }
+                    calleeReference.accept(visitor, data)
+                }
+            }
+
+            override fun visitBlock(block: FirBlock, data: MiniCfgData) {
+                data.variableDeclarations.addLast(mutableMapOf())
+                super.visitBlock(block, data)
+                data.variableDeclarations.removeLast()
+            }
+
+            override fun visitProperty(property: FirProperty, data: MiniCfgData) {
+                if (property.isLocal) {
+                    data.variableDeclarations.last()[property.name] = property
+                }
+            }
+
+            override fun visitVariableAssignment(variableAssignment: FirVariableAssignment, data: MiniCfgData) {
+                val flow = data.flow ?: return
+                val name = (variableAssignment.lValue as? FirNamedReference)?.name ?: return
+                flow.recordAssignment(name, data)
+            }
+
+            override fun visitAssignmentOperatorStatement(assignmentOperatorStatement: FirAssignmentOperatorStatement, data: MiniCfgData) {
+                val flow = data.flow ?: return
+                val lhs = assignmentOperatorStatement.leftArgument as? FirQualifiedAccessExpression ?: return
+                if (lhs.explicitReceiver != null) return
+                val name = (lhs.calleeReference as? FirNamedReference)?.name ?: return
+                flow.recordAssignment(name, data)
+            }
+
+            fun MiniFlow.recordAssignment(name: Name, data: MiniCfgData) {
+                val property = data.resolveLocalVariable(name) ?: return
+                recordAssignment(property, mutableSetOf())
+            }
+
+            private fun MiniFlow.recordAssignment(property: FirProperty, visited: MutableSet<MiniFlow>) {
+                if (this in visited) return
+                visited += this
+                assignedLocalVariables += property
+                parents.forEach { it.recordAssignment(property, visited) }
```

Comment:
Nit: add a comment to explain that assigned local variables are back-propagated here?